sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Expression.scala

```diff
@@ -353,7 +353,7 @@ abstract class BinaryExpression extends Expression with trees.BinaryNode[Express
  * 2. Two inputs are expected to the be same type. If the two inputs have different types,
  *    the analyzer will find the tightest common type and do the proper type casting.
  */
-abstract class BinaryOperator extends BinaryExpression with ExpectsInputTypes {
```

Comment:
We have a rule in `HiveTypeCoercion` for `BinaryOperator`, extending `ExpectsInputTypes` only help us generate error message here. I think it's better to generate a different error message for `BinaryOperator`.